File: ruby/small_problems/easy2.rb
# puts "Teddy is #{rand(20..200)} years old!"

# puts "Enter the length of the room in meters:"
# length = gets.chomp.to_f

# puts "Enter the width of the room in meters:"
# width = gets.chomp.to_f

# square = length * width

# puts "The area of the room is #{square.round 1} square meters (#{(square * 10.7639).round 2} square feet)."

# puts "What is the bill?"
# bill = gets.chomp.to_f
# puts "What is the tip percentage?"
# tip_percent = gets.chomp.to_f
# tip = bill * (tip_percent/100)
# total = bill + tip

# puts "The tip is $#{tip.round 2}"
# puts "The total is $#{total.round 2}"

# require 'date'
# current_year = Date.today.year

# puts "What is your age?"
# current_age = gets.chomp.to_i

# puts "At what age would you like to retire?"
# retirement_age = gets.chomp.to_i

# years_till_retirement = retirement_age - current_age
# retirement_year = current_year + years_till_retirement

# puts "It's #{current_year}. You will retire in #{retirement_year}."
# puts "You have only #{years_till_retirement} years of work to go!"

# puts "What is your name?"
# name = gets.chomp

# if name[-1] == '!'
#     puts "HELLO #{name.gsub('!','').upcase}. WHY ARE WE SCREAMING?"
# else
#     puts "Hello #{name}."
# end

# [*1..99].each { |n| puts n if n.odd? }
# [*1..99].each { |n| puts n if n.even? }

# puts " Please enter an integer greater than 0:"
# int = gets.to_i

# puts "Enter 's' to compute the sum, 'p' to compute the product."
# option = gets.chomp

# if option == 's'
#     sum = [*1..int].sum
#     puts "The sum of the integers between 1 and #{int} is #{sum}."
# end

# if option == 'p'
#     product = [*1..int].reduce(:*)
#     puts "The product of the integers between 1 and #{int} is #{product}."
# end

# name = 'Bob'
# save_name = name
# name.upcase!
# puts name, save_name

def negative(n)
    return n if n < 0
    n * -1
end

puts negative(5) == -5
puts negative(-3) == -3
puts negative(0) == 0      # There's no such thing as -0 in ruby

File: python/small_problems/easy_2.py
def greetings(list, dic):
    name = " ".join(list)
    title = dic['title'] + ' ' + dic['occupation']
    return f"Hello, {name}! Nice to have a {title} around."

greeting = greetings(
    ["John", "Q", "Doe"],
    {"title": "Master", "occupation": "Plumber"},
)
print(greeting)
# Hello, John Q Doe! Nice to have a Master Plumber around.


# name = input('What is your name?' )
# last_char = name[-1]

# if last_char == '!': # name.endswith('!')
#     print(f"HELLO {name.upper()}! WHY ARE WE YELLING?")
# else:
#     print(f"Hello {name}")

def multiply(x, y):
    return x * y

print(multiply(5, 3) == 15)  # True

def square(n):
    return multiply(n, n)

print(square(5) == 25)   # True
print(square(-8) == 64)  # True

# x = float(input('Enter the first number: '))
# y = float(input('Enter the second number: '))

# print(x + y)
# print(x - y)
# print(x * y)
# print(x / y)
# print(x // y)
# print(x % y)
# print(x ** y)

def penultimate(str):
    list = str.split()
    return list[-2]
# These examples should print True
print(penultimate("last word") == "last")
print(penultimate("Launch School is great!") == "is")

def xor(a, b):
    return bool((not (a and b)) and (a or b))

print('xor')
print(xor(5, 0) == True)
print(xor(False, True) == True)
print(xor(1, 1) == False)
print(xor(True, True) == False)

def oddities(list):
    # oddities = []
    # for i in range(0, len(list), 2):
    #     oddities.append(list[i])
    # return oddities
    return list[::2]

print(oddities([2, 3, 4, 5, 6]) == [2, 4, 6])  # True
print(oddities([1, 2, 3, 4]) == [1, 3])        # True
print(oddities(["abc", "def"]) == ['abc'])     # True
print(oddities([123]) == [123])                # True
print(oddities([]) == [])                      # True

from random import randint

print(f"Teddy is {randint(20, 100)} years old!")

# from datetime import date

# current_age = int(input('What is your age? '))
# retirement_age = int(input('At what age would you like to retire? '))
# remaining_years = retirement_age - current_age
# current_year = date.today().year
# retirement_year = current_year + remaining_years

# print(f"It's {current_year}. You will retire in {retirement_year}.")
# print(f"You have only {remaining_years} years of work to go!")

def center_of(s):
    length = len(s)
    middle = length // 2
    center = s[middle:middle + 1] if length % 2 == 1 else s[middle - 1:middle + 1]
    return center

print(center_of('I Love Python!!!') == "Py")    # True
print(center_of('Launch School') == " ")        # True
print(center_of('Launchschool') == "hs")        # True
print(center_of('Launch') == "un")              # True
print(center_of('Launch School is #1') == "h")  # True
print(center_of('x') == "x")                    # True

def negative(n):
    # return abs(n) * -1
    return -abs(n)

print(negative(5) == -5)      # True
print(negative(-3) == -3)     # True
print(negative(0) == 0)       # True